[BUG] `@mandatory` annotation on `function` parameters causes "Unexpected mandatory after optional parameter" error

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current Behavior

When defining a CDS `function` with multiple parameters, all marked as `@mandatory`, the OpenAPI compiler fails with:

```
[ERROR] Unexpected mandatory after optional parameter (in function:"<ServiceName>.<functionName>"/param:"<secondParam>")
```

The compiler appears to not recognize `@mandatory` annotations on `function` parameters at all — every parameter is treated internally as optional, which then triggers the OData V4 rule that mandatory parameters must come before optional ones as soon as a second mandatory parameter is encountered.

The same annotation pattern works correctly on `action` definitions.

### Expected Behavior

The `function` definition compiles successfully and the generated OpenAPI document marks both parameters as `required: true`, identical to how the `action` variant is handled.

### Steps To Reproduce

Minimal CDS service definition:

```cds
service ReproService {

  // FAILS with `odata-parameter-order`
  // Both params are tagged @mandatory, so a developer reasonably expects this to be valid.
  // The first param is an enum type (no `not null`) -> compiler treats it as optional.
  // The second param references a key field (`name` is `String not null`) -> compiler treats it as mandatory.
  // -> "Unexpected mandatory after optional parameter".
  function getMapping(
    @mandatory
    @Core.Description: 'First parameter'
    paramA: TypeA,

    @mandatory
    @Core.Description: 'Second parameter'
    paramB: TypeB:name
  ) returns String;

  // WORKS — identical signature, only `function` -> `action`
  action getMappingAsAction(
    @mandatory
    @Core.Description: 'First parameter'
    paramA: String,

    @mandatory
    @Core.Description: 'Second parameter'
    paramB: String
  ) returns String;
}

define type TypeA: String enum {
  A; B; C;
}

entity TypeB{
  key id : UUID;
  name   : String not null;
}
```

### Environment

```markdown
- `@sap/cds`: 9.3.x
- `@cap-js/openapi`: latest
- Node.js: 24.x
- OS: Windows 11
```

### Repository Containing a Minimal Reproducible Example

_No response_

### Anything else?

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/).
 ### Removed
 ### Fixed
 - Restore bound actions/functions on containment navigation paths (regression since v1.2.0)
+- Function parameters annotated with `@mandatory` no longer lead to "Unexpected mandatory after optional parameter" error
 ### Security
 
 ## [1.4.2] - 2026-05-18

--- a/lib/compile/index.js
+++ b/lib/compile/index.js
@@ -241,10 +241,10 @@ function _readConfigFile(configFilePath) {
  */
 function _applyMandatoryToFunctionParams(csn) {
   const functions = Object.values(csn.definitions ?? {})
-    .filter(def => def.kind === 'function')
+    .filter(def => def.kind === 'function');
   for (const f of functions) {
     for (const p of Object.values(f.params ?? {})) {
-      p.notNull ??= p['@mandatory'] === true
+      p.notNull ??= p['@mandatory'] === true;
     }
   }
 }

--- a/lib/compile/index.js
+++ b/lib/compile/index.js
@@ -39,6 +39,8 @@ function processor(csn, options = {}) {
   , ...options
 }
 
+  _applyMandatoryToFunctionParams(csn);
+
   // must not be part of function* otherwise thrown errors are swallowed
   const csdl = cds.compile.to.edm(csn, edmOptions);
   let openApiDocs = {};
@@ -227,6 +229,24 @@ function _readConfigFile(configFilePath) {
     result[normalizedKey] = typeof value === 'object' && value !== null ? JSON.stringify(value) : value;
   }
   return result;
+}
+
+/**
+ * The cds-compiler's edmPreprocessor determines function parameter optionality
+ * from the type's nullability (p.notNull), ignoring the CDS `@mandatory` annotation.
+ * A nullable-typed parameter annotated @mandatory` is incorrectly treated as optional,
+ * which then triggers an odata-parameter-order error if a non-nullable parameter follows.
+ *
+ * Fix: honour `@mandatory` by setting notNull on the CSN param before EDM compilation.
+ */
+function _applyMandatoryToFunctionParams(csn) {
+  const functions = Object.values(csn.definitions ?? {})
+    .filter(def => def.kind === 'function')
+  for (const f of functions) {
+    for (const p of Object.values(f.params ?? {})) {
+      p.notNull ??= p['@mandatory'] === true
+    }
+  }
 }
 
 // we're attaching the events to the main function so they become automatically exposed through the cds facade:

--- a/test/lib/compile/openapi.test.js
+++ b/test/lib/compile/openapi.test.js
@@ -684,4 +684,22 @@ service CatalogService {
       cds.removeListener(events.after, handler);
     }
   });
+
+  test('@mandatory on function parameters with nullable types compiles without error', () => {
+    const csn = cds.compile.to.csn(`
+      service ReproService {
+        function getMapping(
+          @mandatory paramA: String,
+          @mandatory paramB: String not null
+        ) returns String;
+      }
+    `);
+    const openAPI = toOpenApi(csn);
+    const params = openAPI.paths['/getMapping'].get.parameters;
+    assert.strictEqual(params.length, 2);
+    assert.strictEqual(params[0].name, 'paramA');
+    assert.strictEqual(params[0].required, true);
+    assert.strictEqual(params[1].name, 'paramB');
+    assert.strictEqual(params[1].required, true);
+  });
 });

